File: src/main/java/com/github/fherbreteau/gatling/ftp/javaapi/Ftp.java
package com.github.fherbreteau.gatling.ftp.javaapi;

import com.github.fherbreteau.gatling.ftp.javaapi.action.FtpActionBuilder;
import io.gatling.commons.validation.Validation;
import io.gatling.core.session.Session;
import io.gatling.javaapi.core.internal.Expressions;
import scala.Function1;

import javax.annotation.Nonnull;
import java.util.function.Function;

public class Ftp {

    private final Function1<Session, Validation<String>> name;

    Ftp(Function1<io.gatling.core.session.Session, Validation<String>> name) {
        this.name = name;
    }

    public FtpActionBuilder move(@Nonnull String file) {
        return new FtpActionBuilder(new com.github.fherbreteau.gatling.ftp.Ftp(name).move(Expressions.toStringExpression(file)));
    }

    public FtpActionBuilder move(@Nonnull Function<io.gatling.javaapi.core.Session, String> file) {
        return new FtpActionBuilder(new com.github.fherbreteau.gatling.ftp.Ftp(name).move(Expressions.javaFunctionToExpression(file)));
    }

    public FtpActionBuilder copy(@Nonnull String file) {
        return new FtpActionBuilder(new com.github.fherbreteau.gatling.ftp.Ftp(name).copy(Expressions.toStringExpression(file)));
    }

    public FtpActionBuilder copy(@Nonnull Function<io.gatling.javaapi.core.Session, String> file) {
        return new FtpActionBuilder(new com.github.fherbreteau.gatling.ftp.Ftp(name).copy(Expressions.javaFunctionToExpression(file)));
    }

    public FtpActionBuilder delete(@Nonnull String file) {
        return new FtpActionBuilder(new com.github.fherbreteau.gatling.ftp.Ftp(name).delete(Expressions.toStringExpression(file)));
    }

    public FtpActionBuilder delete(@Nonnull Function<io.gatling.javaapi.core.Session, String> file) {
        return new FtpActionBuilder(new com.github.fherbreteau.gatling.ftp.Ftp(name).delete(Expressions.javaFunctionToExpression(file)));
    }

    public FtpActionBuilder upload(@Nonnull String file) {
        return new FtpActionBuilder(new com.github.fherbreteau.gatling.ftp.Ftp(name).upload(Expressions.toStringExpression(file)));
    }

    public FtpActionBuilder upload(@Nonnull Function<io.gatling.javaapi.core.Session, String> file) {
        return new FtpActionBuilder(new com.github.fherbreteau.gatling.ftp.Ftp(name).upload(Expressions.javaFunctionToExpression(file)));
    }

    public FtpActionBuilder download(@Nonnull String file) {
        return new FtpActionBuilder(new com.github.fherbreteau.gatling.ftp.Ftp(name).download(Expressions.toStringExpression(file)));
    }

    public FtpActionBuilder download(@Nonnull Function<io.gatling.javaapi.core.Session, String> file) {
        return new FtpActionBuilder(new com.github.fherbreteau.gatling.ftp.Ftp(name).download(Expressions.javaFunctionToExpression(file)));
    }
}

File: src/main/scala/io/github/fherbreteau/gatling/ftp/Ftp.scala
package io.github.fherbreteau.gatling.ftp

import io.gatling.core.session.{EmptyStringExpressionSuccess, Expression}
import io.github.fherbreteau.gatling.ftp
import io.github.fherbreteau.gatling.ftp.action.FtpActionBuilder
import io.github.fherbreteau.gatling.ftp.client.FtpActions._

case class Ftp(operationName: Expression[String]) {

  def mkdir(directory: Expression[String]): FtpActionBuilder = action(directory, EmptyStringExpressionSuccess, Mkdir)

  def move(source: Expression[String], destination: Expression[String]): FtpActionBuilder = action(source, destination, Move)

  def copy(source: Expression[String], destination: Expression[String]): FtpActionBuilder = action(source, destination, Copy)

  def download(file: Expression[String]): FtpActionBuilder = download(file, file)

  def download(source: Expression[String], destination: Expression[String]): FtpActionBuilder = action(source, destination, Download)

  def upload(file: Expression[String]): FtpActionBuilder = upload(file, file)

  def upload(source: Expression[String], destination: Expression[String]): FtpActionBuilder = action(source, destination, Upload)

  def delete(file: Expression[String]): FtpActionBuilder = action(file, EmptyStringExpressionSuccess, Delete)

  def rmdir(directory: Expression[String]): FtpActionBuilder = action(directory, EmptyStringExpressionSuccess, RmDir)

  private def action(source: Expression[String], destination: Expression[String], action: Action): FtpActionBuilder =
    FtpActionBuilder(operationName, source, destination, action)
}
